src/chainstate/stacks/boot/cost-voting.clar

```diff
@@ -0,0 +1,208 @@
+                                function-name: function-name,
+                                expiration-block-height: (+ burn-block-height u2016) })
+        (map-insert proposal-votes { proposal-id: (var-get proposal-count) } { votes: u0 })
+        (var-set proposal-count (+ (var-get proposal-count) u1))
+        (ok (- (var-get proposal-count) u1))))
+
+;; Vote on a proposal
+(define-public (vote-proposal (proposal-id uint) (amount uint))
+    (let (
+        (cur-votes (default-to u0 (get votes (map-get? proposal-votes { proposal-id: proposal-id }))))
+        (cur-principal-votes (default-to u0 (get votes (map-get? principal-proposal-votes {
+            address: tx-sender,
+            proposal-id: proposal-id }))))
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err 1))))
+    )
+    (if (and
+            (< burn-block-height expiration-block-height)
+            (is-none (map-get? proposal-confirmed-id { proposal-id: proposal-id }))
+        )
+        (begin
+            (unwrap! (stx-transfer? amount tx-sender (as-contract tx-sender)) (err 1))
+            (unwrap! (ft-mint? cost-vote-token amount tx-sender) (err 1))
+            (map-set proposal-votes { proposal-id: proposal-id } { votes: (+ amount cur-votes) })
+            (map-set principal-proposal-votes { address: tx-sender, proposal-id: proposal-id}
+                                            { votes: (+ amount cur-principal-votes)})
+            (ok true))
+    (err 1)))
+)
+
+;; Withdraw votes
+(define-public (withdraw-votes (proposal-id uint) (amount uint))
+    (let (
+        (cur-votes (default-to u0 (get votes (map-get? proposal-votes { proposal-id: proposal-id }))))
+        (cur-principal-votes (default-to u0 (get votes (map-get? principal-proposal-votes {
+            address: tx-sender,
+            proposal-id: proposal-id }))))
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err 1))))
+        (sender tx-sender)
+    )
+    (if (and
+            (>= cur-principal-votes amount)
+            (< burn-block-height expiration-block-height)
+            ;;(is-none (map-get? confirmed-proposals { proposal-id: proposal-id }))
+        )
+        (begin
+            (unwrap! (as-contract (stx-transfer? amount tx-sender sender)) (err 1))
+            (unwrap! (as-contract (ft-transfer? cost-vote-token amount sender tx-sender)) (err 1))
+            (map-set proposal-votes { proposal-id: proposal-id } { votes: (- cur-votes amount) })
+            (map-set principal-proposal-votes { address: tx-sender, proposal-id: proposal-id }
+                                              { votes: (- cur-principal-votes amount) })
+            (ok true))
+        (err 1)))
+)
+
+;; Withdraw STX after vote is over
+(define-public (withdraw-after-votes (proposal-id uint))
```

Comment:
Is `withdraw-after-votes` necessary? Could a voter just use `withdraw-votes` if we removed the `expiration-block-height` check?

Reply:
I don't think it's necessary. Will make this change. I think we'll want `withdraw-votes` to behave slightly differently if the proposal is expired or not though. If the proposal is expired, we can keep the vote count the same (to preserve a record of the vote)